docs/ADR-10-cardano-api-script-witness-api.md

```diff
@@ -0,0 +1,217 @@
+  [ (ScriptWitnessIndexMint ix, policyId', assetName', quantity, witness)
+  | (ix, (policyId', assets)) <- zip [0 ..] $ toList policiesWithAssets
+  , (assetName', quantity, witness) <- assets
+  ]
+
+...
+
+indexTxIns
+  :: TxIns BuildTx era
+  -> [(ScriptWitnessIndex, TxIn, Witness WitCtxTxIn era)]
+indexTxIns txins =
+  [ (ScriptWitnessIndexTxIn ix, txIn, witness)
+  | (ix, (txIn, BuildTxWith witness)) <- zip [0 ..] $ orderTxIns txins
+  ]
+
+...etc
+
+```
+
+There is a repeating pattern here that basically has something that is _witnessable_ (tx input, mint value, certificate, etc) accompanied by its index and the corresponding witness. Therefore the following types are introduced to capture this pattern:
+
+```haskell
+data Witnessable thing era where
+  WitTxIn :: L.AlonzoEraScript era => TxIn -> Witnessable TxIn era
+  WitTxCert :: L.AlonzoEraScript era => Cert -> Witnessable Cert era
+  WitMint :: L.AlonzoEraScript era => Mint -> Witnessable Mint era
+  WitWithdrawal
+    :: L.AlonzoEraScript era => Withdrawal -> Witnessable Withdrawal era
+  WitVote
+    :: L.ConwayEraScript era
+    => Voter -> Witnessable Voter era
+  WitProposal :: L.ConwayEraScript era => Proposal -> Witnessable Proposal era
+
+data AnyWitness era where
+  AnyKeyWitness :: AnyWitness era
+  AnySimpleScriptWitness :: SimpleScriptOrReferenceInput era -> AnyWitness era
+  AnyPlutusScriptWitness :: PlutusScriptWitness lang purpose era -> AnyWitness era
+```
+
+And the upshot is once we have `[(Witnessable thing era, AnyWitness era)]` these can be uniformly treated to construct the redeemer pointer map (`Redeemers era`):
+
+```haskell
+
+data TxScriptWitnessRequirements era
+  = TxScriptWitnessRequirements
+      (Set L.Language)
+      [L.Script era]
+      (L.TxDats era)
+      (L.Redeemers era) -- Redeemer pointer map
+
+getTxScriptWitnessesRequirements
+  :: AlonzoEraOnwards era
+  -> [(Witnessable witnessable (ShelleyLedgerEra era), AnyWitness (ShelleyLedgerEra era))]
+  -> TxScriptWitnessRequirements (ShelleyLedgerEra era)
+getTxScriptWitnessesRequirements eon wits =
+  obtainMonoidConstraint eon $ mconcat $ map (getTxScriptWitnessRequirements eon) wits
+```
+
+## New `IndexedPlutusScriptWitness` types
+
+We are only concerned with the index of something that is plutus script witnessed. Therefore we introduce `IndexedPlutusScriptWitness`:
+
+```haskell
+data IndexedPlutusScriptWitness witnessable (lang :: L.Language) (purpose :: PlutusScriptPurpose) era where
+  IndexedPlutusScriptWitness
+    :: Witnessable witnessable era
+    -> (L.PlutusPurpose L.AsIx era)
+    -> (PlutusScriptWitness lang purpose era)
+    -> IndexedPlutusScriptWitness witnessable lang purpose era
+```
+
+We do away with `cardano-api`'s `ScriptWitnessIndex` and directly use `cardano-ledger`'s `PlutusPurpose AsIx era` which is essentially the same thing. 
+The benefit here is we get to use ledger type classes to calculate the index of the plutus script witness thing we are interested in:
```

Comment:
>calculate

I'm not sure if that's the right word here for `Word32 -> PlutusPurpose AsIx era`. We're not really calculating the index here, but just wrapping it correctly. Maybe "obtain" or "build" would work here better?

Reply:
Went with produce